### DISCUSSION TITLE
Zoom status API details question

### DISCUSSION THREAD

**Discussion by itspamlam** [digirati-co-uk/iiif-canvas-panel#231] [Q&A]
In our implementation of Canvas Panel, we added buttons to control zooming in & out. We added a `zoom` event listener to help determine whether or not to disable either button by checking both `isMin` & `isMax`. However, on occasion, I've noticed that `isMin` has not flipped to `true` when I would have thought it would. When I check the event details in those cases, I noticed that the current scale / scaleFactor is actually less than `min` which I would not have expected. The below screenshot is after having zoomed in once - I think I would have expected `min` to be `0`?

<img width="419" alt="Screenshot 2023-12-20 at 3 43 08 PM" src="https://github.com/digirati-co-uk/iiif-canvas-panel/assets/25655259/4d541982-166b-4fa1-9f5a-8c6a12291802">

When I've used OpenSeadragon, there have been times when I would need to set their `minZoomImageRatio` or `minZoomLevel` - is there any equivalent that I can/should be using here?

**Comment by stephenwf** [comment]
Firstly, it looks like `min` value there might be a bug, but `isMin` looks correct. (If you zoom in, you should be able to zoom back out). 

However, I think it might be useful to describe how the zoom works, and how it might be different to other deep zoom viewers, and that might help in adding an extra options to canvas panel - or using the existing data from the event. 

The zoom is the ratio between the real pixels and what is displayed. So a 3000px at a scale of 1 would be zoomed in completely. By default the max is 1 (so pixels match), and the minimum is the image scaled to fit the viewer. So the minimum (and value of `scaleFactor`) should always be greater than zero. 

The underlying viewer component does have 3 options to change these values.
```js
visibilityRatio: number;
maxOverZoom: number;
maxUnderZoom: number;
```
https://github.com/atlas-viewer/atlas/blob/feature/updates/src/renderer/runtime.ts#L31-L35

We could add attributes to canvas panel and pass these values down:
https://github.com/digirati-co-uk/iiif-canvas-panel/blob/ae2a4ac2c4a2d64578d1d8631772925cd532c3c1/packages/canvas-panel/src/hooks/use-generic-atlas-props.ts#L603

The visibility ratio is how much you can "move" the image off the side of the screen. 

![image](https://github.com/digirati-co-uk/iiif-canvas-panel/assets/8266711/e387dfec-6147-442c-8b96-2721832d1c4e)
![image](https://github.com/digirati-co-uk/iiif-canvas-panel/assets/8266711/7639903d-ed89-4bb7-949b-a2b03ad8ce72)


The two examples above show the default "50%" value, where at least half of the image must be shown (it does this for x and y separately, see below)

![image](https://github.com/digirati-co-uk/iiif-canvas-panel/assets/8266711/926e3acb-5d18-4d3e-941b-6e49a57ddb24)


### Zoom levels

Canvas panel doesn't support a fixed max or min zoom level, as this will change depending on the image(s) that are being displayed. Instead it allows you to change the "over" and "under" zoom levels. `maxOverZoom` will set an "over-zoom" functionality, so the image quality will decrease but you can zoom further (e.g. a value of 2 would let you zoom 2x so each pixel was 2x2 pixels on the screen). The `maxUnderZoom` is how far you can zoom out past fitting the viewport, so a value of `0.5` would let you scale the image to half the size of the viewport, regardless of its size. However this conflicts with the previously mentioned `visibilityRatio` and can cause some strange bugs.

**Comment by itspamlam** [comment]
Thanks @stephenwf - the detailed explanation of how zoom works was helpful, I'll make note of it for our team's reference. I can imagine there's a determination of how much detail should be included in the docs, but adding some of this could be nice.

I just played around with the default behaviors for the zoom levels & for our initial goal of just swapping out our current viewers with canvas panel I don't think we'd want to touch any of those initially, especially if it could decrease the image quality - that would defeat one of the purposes of switching to canvas panel. But as we build out additional features we'll try and remember those are possibilities if applicable.

However, going back to the `min` issue, if it is a bug does an issue need to be opened up for it? In general, I was just trying to get my questions in before the holidays. Thanks for being so responsive & happy holidays!

**Reply by stephenwf** [reply]
Yes, I think we could create a new bug issue for the min value. If you have some more details about the size of image and size of the viewer - that would help debug the issue.

Thanks, Happy holidays!

**Reply by itspamlam** [reply]
Of course -
manifest-url: https://media.getty.edu/iiif/manifest/6a744965-6345-41cf-8885-69dd07e25008
canvas height: 500
